network/src/protocols/test.rs

```diff
@@ -0,0 +1,386 @@
+use super::{
+    discovery::{DiscoveryProtocol, DiscoveryService},
+    feeler::Feeler,
+    identify::IdentifyCallback,
+    ping::PingService,
+};
+
+use crate::{
+    network::EventHandler,
+    network::{DISCOVERY_PROTOCOL_ID, FEELER_PROTOCOL_ID, IDENTIFY_PROTOCOL_ID, PING_PROTOCOL_ID},
+    NetworkConfig, NetworkState,
+};
+
+use std::{sync::Arc, thread, time::Duration};
+
+use futures::{
+    sync::mpsc::{self, channel},
+    Stream,
+};
+use p2p::{
+    builder::{MetaBuilder, ServiceBuilder},
+    multiaddr::Multiaddr,
+    service::{DialProtocol, ProtocolHandle, ServiceControl, TargetProtocol},
+    utils::multiaddr_to_socketaddr,
+    ProtocolId, SessionId,
+};
+use p2p_identify::IdentifyProtocol;
+use p2p_ping::PingHandler;
+use tempfile::tempdir;
+
+struct Node {
+    listen_addr: Multiaddr,
+    control: ServiceControl,
+    network_state: Arc<NetworkState>,
+}
+
+impl Node {
+    fn dial(&self, node: &Node, protocol: DialProtocol) {
+        let _ = self.control.dial(node.listen_addr.clone(), protocol);
```

Comment:
If we assume that dial will be ok, then can we use `unwrap` here?
If not, ignore my comment.
The same below.